core/primitives/src/bandwidth_scheduler.rs

```diff
@@ -79,6 +79,54 @@ pub struct BandwidthRequest {
     pub requested_values_bitmap: BandwidthRequestBitmap,
 }
 
+impl BandwidthRequest {
+    /// Creates a bandwidth request based on the sizes of receipts in the outgoing buffer.
+    /// Returns None when a request is not needed (receipt size below base bandwidth).
+    pub fn make_from_receipt_sizes(
+        to_shard: u8,
+        receipt_sizes: impl Iterator<Item = u64>,
+        params: &BandwidthSchedulerParams,
+    ) -> Option<BandwidthRequest> {
+        let values = BandwidthRequestValues::new(params).values;
+        let mut bitmap = BandwidthRequestBitmap::new();
+
+        // For every receipt find out how much bandwidth would be needed to send out
+        // all the receipts up to this one. Then find the value that is at least as
+        // large as the required bandwidth and request it in the request bitmap.
+        let mut total_size: u64 = 0;
+        let mut cur_value_idx: usize = 0;
+        for receipt_size in receipt_sizes {
+            total_size = total_size.checked_add(receipt_size).expect(
+                "Total size of receipts doesn't fit in u64, are there exabytes of receipts?",
+            );
+
+            if total_size <= params.base_bandwidth {
+                continue;
+            }
+
+            // Find a value that is at least as big as the total_size
+            while cur_value_idx < values.len() && values[cur_value_idx] < total_size {
+                cur_value_idx += 1;
+            }
+
+            if cur_value_idx == values.len() {
+                // There is no value to request this much, stop the loop.
+                break;
+            }
```

Comment:
Should you also request the largest value (set the last bit) in this case? I'm not sure myself, just checking.

Reply:
I initially did that, but I think it's the wrong thing to do.
Having some `total_size` means that we have the ability to send this much, not less and not more. It doesn't make sense to request less bandwidth than what we can send.
The `two_max_size_receipts` test is a good example - we want to send two max size receipts. It doesn't make sense to request `max_shard_bandwidth` as it doesn't allow us to send two receipts.